### DISCUSSION TITLE
brew install python3 installed python 3.10 despite that python 3.11 is available? Why does Heombrew do that? How can I make sure I'm always using the newest version of python3?

### DISCUSSION THREAD

**Discussion by YousufSSyed** [Homebrew/discussions#4056] [Everyday usage] [answered]
I just did `brew install python@3.11`, but does that pin it to that specific version? I wouldn't want to have to check occasionally if I'm on the latest Python version. Also, when I uninstall the version of Python already installed with homebrew and install 3.11, entering in `python3` invokes macOS' python, and after restarting the terminal. How could I replace Python 3.10 (installed wit Homebrew) and drop in 3.11?

**Comment by p-linnane** [comment, marked as answer]
[`python`](https://github.com/Homebrew/homebrew-core/blob/master/Aliases/python), [`python3`](https://github.com/Homebrew/homebrew-core/blob/master/Aliases/python3), and [`python@3`](https://github.com/Homebrew/homebrew-core/blob/master/Aliases/python%403) are all currently aliased to `python@3.10`. We are in the middle of a migration from 3.10 to 3.11.

Did you run `brew install python@3.11`? Or one of the aliases above?

**Reply by YousufSSyed** [reply]
Did u read the post? I did do `brew install python@3.11`.

**Reply by p-linnane** [reply]
I did read the post, and I asked some clarifying questions to confirm. 

If you specifically installed 3.11, that is what you will stay on for at least the next year. When 3.12 comes out next fall, we will go through another migration process from 3.11 to 3.12.

**Reply by YousufSSyed** [reply]
I did. I get the Pyhton 3.11 interpreter when I enter in python3.11, but I'd like that to be the default, like when I type in `python3`; and would it be possible to have all my pip packages use that Python version instead? How long does it usually take for Homebrew to migrate versions?

**Reply by SMillerDev** [reply]
The Python install prints instructions on how to ensure that particular python is used as `python3`.

**Reply by YousufSSyed** [reply]
Are these the instructions? It outputted after doing `brew reinstall python@3.11`

```
Unversioned and major-versioned symlinks `python`, `python3`, `python-config`, `python3-config`, `pip`, `pip3`, etc. pointing to
`python3.11`, `python3.11-config`, `pip3.11` etc., respectively, have been installed into
  /opt/homebrew/opt/python@3.11/libexec/bin

If you do not need a specific version of Python, and always want Homebrew's `python3` in your PATH:
  brew install python3

You can install Python packages with
  pip3.11 install <package>
They will install into the site-package directory
  /opt/homebrew/lib/python3.11/site-packages

tkinter is no longer included with this formula, but it is available separately:
  brew install python-tk@3.11

gdbm (`dbm.gnu`) is no longer included in this formula, but it is available separately:
  brew install python-gdbm@3.11
`dbm.ndbm` changed database backends in Homebrew Python 3.11.
If you need to read a database from a previous Homebrew Python created via `dbm.ndbm`,
you'll need to read your database using the older version of Homebrew Python and convert to another format.
`dbm` still defaults to `dbm.gnu` when it is installed.

For more information about Homebrew and Python, see: https://docs.brew.sh/Homebrew-and-Python
```

It doesn't say exactly what to do.

**Reply by gromgit** [reply]
>```
> Unversioned and major-versioned symlinks `python`, `python3`, `python-config`, `python3-config`, `pip`, `pip3`, etc. pointing to
> `python3.11`, `python3.11-config`, `pip3.11` etc., respectively, have been installed into
>   /opt/homebrew/opt/python@3.11/libexec/bin
>```

This says that symbolic links (the "aliases" that @p-linnane mentioned previously) to the relevant `python@3.11` binaries, with major-version-only (`python3`, etc.) and _no-version_ (`python`, etc.) names have been placed in `/opt/homebrew/opt/python@3.11/libexec/bin`.

If that still doesn't make any sense to you, look in `/opt/homebrew/opt/python@3.11/libexec/bin`.

As to how to use those symlinks, you have two options:
* run them directly with full paths, e.g. `/opt/homebrew/opt/python@3.11/libexec/bin/python3`
* add `/opt/homebrew/opt/python@3.11/libexec/bin` to your `PATH`, then you can just run `python3`

**Comment by marcelstoer** [comment]
I am fine with 3.10 being the default. However, I find it odd that installing `virtualenv` then installs 3.11.

```
$ virtualenv --version
virtualenv 20.17.1 from /usr/local/Cellar/virtualenv/20.17.1/libexec/lib/python3.11/site-packages/virtualenv/__init__.py
```
~~Haven't figured out yet which `virtualenv` I would need such that it (also) uses 3.10.~~

It appears 20.16.7 is the last version that [depends on Python 3.10](https://github.com/Homebrew/homebrew-core/blob/ad8df1de005bba0b0321789c6a1a0091e8cc8da0/Formula/virtualenv.rb#L22).
